certification/internal/k8s/deployable_by_olm.go

```diff
@@ -0,0 +1,242 @@
+		return err
+	}
+	key = "package"
+	packageName, err = viperutil.GetString(key)
+	if err != nil {
+		log.Error(fmt.Sprintf("To set the key, export PFLT_%s or add %s:<value> to config.yaml in the current working directory", strings.ToUpper(key), key))
+		return err
+	}
+	key = "app"
+	app, err = viperutil.GetString(key)
+	if err != nil {
+		log.Error(fmt.Sprintf("To set the key, export PFLT_%s or add %s:<value> to config.yaml in the current working directory", strings.ToUpper(key), key))
+		return err
+	}
+	key = "namespace"
+	ooNamespace, err = viperutil.GetString(key)
+	if err != nil {
+		log.Error(fmt.Sprintf("To set the key, export PFLT_%s or add `%s:<value>` to config.yaml in the current working directory", strings.ToUpper(key), key))
+		return err
+	}
+
+	_, err = openshiftEngine.CreateNamespace(ooNamespace, cli.OpenshiftOptions{}, k8sconfig)
+	if err != nil && !k8serrors.IsAlreadyExists(err) {
+		return err
+	}
+
+	_, err = openshiftEngine.CreateCatalogSource(cli.CatalogSourceData{Name: app, Image: catalogImage}, cli.OpenshiftOptions{Namespace: catalogSourceNS}, k8sconfig)
+	if err != nil && !k8serrors.IsAlreadyExists(err) {
+		return err
+	}
+
+	_, err = openshiftEngine.CreateOperatorGroup(cli.OperatorGroupData{Name: app, TargetNamespaces: targetNamespaces}, cli.OpenshiftOptions{Namespace: ooNamespace}, k8sconfig)
+	if err != nil && !k8serrors.IsAlreadyExists(err) {
+		return err
+	}
+
+	subscriptionData := cli.SubscriptionData{
+		Name:                   app,
+		Channel:                channel,
+		CatalogSource:          app,
+		CatalogSourceNamespace: catalogSourceNS,
+		Package:                packageName,
+	}
+	_, err = openshiftEngine.CreateSubscription(subscriptionData, cli.OpenshiftOptions{Namespace: ooNamespace}, k8sconfig)
+	if err != nil && !k8serrors.IsAlreadyExists(err) {
+		return err
+	}
+	return nil
+}
+
+func (p *DeployableByOlmCheck) isCSVReady(installedCSV string) (bool, error) {
+
+	var csv *operatorv1alpha1.ClusterServiceVersion
+
+	log.Trace(fmt.Sprintf("Looking for csv %s in namespace %s", installedCSV, ooNamespace))
+
+	// query API server for CSV by name
+	ctx, cancel := context.WithTimeout(context.Background(), 90*time.Second)
+	defer cancel()
+
+	for {
+		log.Debug("Waiting for ClusterServiceVersion to become ready...")
+		csv, _ = openshiftEngine.GetCSV(installedCSV, cli.OpenshiftOptions{Namespace: ooNamespace}, k8sconfig)
+		// if the CSV phase is succeeded or the context deadline exceeds, stop the querying
+		if csv.Status.Phase == operatorv1alpha1.CSVPhaseSucceeded {
+			log.Debug("CSV is created successfully: ", installedCSV)
+			return true, nil
+		}
+		// if the context deadline exceeds, fail the check
+		if ctx.Err() != nil {
+			log.Error(fmt.Sprintf("failed to fetch the csv %s: ", installedCSV), ctx.Err())
+			break
+		}
+		log.Debug("CSV is not ready yet, retrying...")
+		time.Sleep(2 * time.Second)
+	}
+	return false, nil
```

Comment:
I think we need a select statement in this block, and any others that are leveraging `context.WithTimeout`.

You're effectively running these operations with a 90 second timeout, but you're checking for the timeout evaluation serially, and I think what you wan to do is similar to what is seen in this example, where you run your logic every N seconds using `time.After`, or fail as soon as the context deadline is exceeded.

https://pkg.go.dev/context#example-WithTimeout

Reply:
Done.. please let me know if you have any feedback :)